File: codewars/6kyu/give-me-a-diamond/dduck-give-me-a-diamond.py
# Give me a Diamond
# https://www.codewars.com/kata/5503013e34137eeeaa001648/

from unittest import TestCase


def diamond(n, count=1, result=''):
    if n % 2 == 0 or n < 1:
        return None
    if n >= count:
        diamondSize = abs(n - abs(n - (2 * count - 1)))
        emptySpaceSize = int((n - diamondSize) / 2)
        emptySpace = " " * emptySpaceSize
        result += emptySpace + ("*" * diamondSize) + "\n"

    if n == count:
        return result

    return diamond(n, count + 1, result)


expected = " *\n"
expected += "***\n"
expected += " *\n"
TestCase().assertEqual(diamond(1), "*\n")
TestCase().assertEqual(diamond(2), None)
TestCase().assertEqual(diamond(3), expected)
TestCase().assertEqual(diamond(5), "  *\n ***\n*****\n ***\n  *\n")
TestCase().assertEqual(diamond(0), None)
TestCase().assertEqual(diamond(-3), None)

File: codewars/6kyu/give-me-a-diamond/dduck-give-me-a-diamond.js
// give-me-a-diamond
// https://www.codewars.com/kata/5503013e34137eeeaa001648/

const diamond = (n, count = 1, result = '') => {
  if (n % 2 === 0 || n < 1) {
    return null;
  }
  if (n >= count) {
    const diamondSize = Math.abs(n - Math.abs(n - (2 * count - 1)));
    const emptySpaceSize = (n - diamondSize) / 2;
    const emptySpace = ' '.repeat(emptySpaceSize);
    result += `${emptySpace}${'*'.repeat(diamondSize)}\n`;
  }
  if (n === count) {
    return result;
  }
  return diamond(n, count + 1, result);
};

console.log(diamond(1), "*\n");
console.log(diamond(3), " *\n***\n *\n");
console.log(diamond(5), "  *\n ***\n*****\n ***\n  *\n");
console.log(diamond(2), null);
console.log(diamond(-3), null);
console.log(diamond(0), null);
